### PR TITLE
OTA-861: Set Upgradeable=False when there is an upgrade in progress

### PR DESCRIPTION
**Pull request by hongkailiu**
This PR add a Upgradeable which fails on Processing=True in `clusterversion.status.conditions`.
In other words, `Upgradeable=False` if an upgrade is in progress, including both minor level and patch level.

In addition, this PR syncs the upgradeable at the shutdown time to ensure the unsaved (due to by the throttle) upgradeable to be saved. This part could be a separate PR too.

For example, it blocks the upgrade to 4.16.1 until the ongoing upgrade
4.14.35 -> 4.15.29 completes.

It also covers the case 4.14.15-> 4.14.35 -> 4.15.29
where the upgrade 4.14.35 -> 4.15.29 is blocked until the upgrade
4.14.15-> 4.14.35 completes.

Note that we still allow for upgrade to 4.y+1.z''
in the middle of upgrade 4.y.z -> 4.y+1.z', even though direct upgrade
4.y.z -> 4.y+1.z'' might not be supported.
This is because the ugprade 4.y.z -> 4.y+1.z' might not be completed
up to a bug in 4.y+1.z' that has a fix in 4.(y+1).z''.
We need the retarget to it to land 4.y+1 on the cluster.

For [OTA-861](https://issues.redhat.com//browse/OTA-861), the guard on retargeting to a minor level upgrade will be added with a follow up PR.

---

Update:
With the throttle of "upgradeable" disabled on the shutdown, the racing window between "CVO starts rolling the new version out" and "CVO gets shutdown" is very short (less than one second in [the test](https://redhat-internal.slack.com/archives/CJ1J9C3V4/p1728060448767569?thread_ts=1727723159.289959&cid=CJ1J9C3V4)).
We do not need to add the second guard back that was dropped in the 362e9ca912edd948b93c2b0545c508cb4bf7bd84.

The acceptedRisks for Y-then-Z upgrade in the dropped commit will be done with another PR.

### CODE REVIEW

**Comment by hongkailiu** [comment]
This PR is replacing https://github.com/openshift/cluster-version-operator/pull/1079

**Comment by wking** [comment]
Exercise a 4.17 -> this-pull update, so we can see `Upgradeable=False` while we're mid-update.

/payload-job periodic-ci-openshift-release-master-ci-4.18-upgrade-from-stable-4.17-e2e-gcp-ovn-upgrade

**Comment by petr-muller** [comment]
/uncc

David and Trevor are involved in this one, seems enough ;)

**Comment by Davoska** [comment]
/title OTA-861: inhibit the 2nd minor version upgrade

**Comment by Davoska** [comment]
/retitle OTA-861: inhibit the 2nd minor version upgrade

File: pkg/cvo/upgradeable.go [hunk 1]
```diff
@@ -291,23 +291,16 @@ func (check *minorVersionClusterUpgradeInProgressUpgradeable) Check() *configv1.
...
-	currentMinor := clusterversion.GetEffectiveMinor(currentVersion)
-	desiredMinor := clusterversion.GetEffectiveMinor(cv.Status.Desired.Version)
-	klog.V(2).Infof("The current minor version is %s and the desired minor version is %s", currentMinor, desiredMinor)
-	if !clusterversion.MinorVersionUpgrade(currentMinor, desiredMinor) {
+	if currentVersion == "" || cv.Status.Desired.Version == "" || currentVersion == cv.Status.Desired.Version {
 		return nil
 	}
 
-	message := fmt.Sprintf("There is a minor level upgrade from %s to %s in progress.", currentVersion, cv.Status.Desired.Version)
+	message := fmt.Sprintf("There is an upgrade from %s to %s in progress.", currentVersion, cv.Status.Desired.Version)
```
Comment (wking) [inline pkg/cvo/upgradeable.go:301]: possibly a more robust way to implement this guard would be to ignore `history` and just echo the `Progressing` status (which is also `True` during installs, and blocking folks from heading to 4.(y+1) before their install completes seems reasonable).

Comment (hongkailiu) [reply pkg/cvo/upgradeable.go:291]: https://github.com/openshift/cluster-version-operator/pull/1080/commits/74d4e0437723ab4bf3e2c46cf20b8e9a2f78b11e to address this.

It seems lots of simpler than before.

File: pkg/payload/precondition/clusterversion/upgradeable.go [hunk 2]
```diff
@@ -66,27 +64,22 @@ func (pf *Upgradeable) Run(ctx context.Context, releaseContext precondition.Rele
 		}
 	}
 
-	if upgradeableUpgradeInProgress := resourcemerge.FindOperatorStatusCondition(cv.Status.Conditions, UpgradeInProgress); upgradeableUpgradeInProgress != nil && upgradeableUpgradeInProgress.Status == configv1.ConditionTrue {
+	if condition := resourcemerge.FindOperatorStatusCondition(cv.Status.Conditions, configv1.OperatorProgressing); condition != nil &&
+		condition.Status == configv1.ConditionTrue &&
+		condition.Reason == ConditionReasonMinorVersionClusterUpgradeInProgress {
```
Comment (wking) [inline pkg/payload/precondition/clusterversion/upgradeable.go:69]: But we don't need this condition, right?  5ea5ebe0e12bf1 is pointing out that we want to be `Upgradeable=False` and blocking 4.y.z' -> 4.(y+1).z" retargets regardless of whether the update we're currently in the middle of is 4.y.z -> 4.y.z' or 4.(y-1).z -> 4.y.z'.

File: pkg/payload/precondition/clusterversion/upgradeable_test.go [hunk 3]
```diff
@@ -114,6 +136,7 @@ func TestUpgradeableRun(t *testing.T) {
 				Spec:       configv1.ClusterVersionSpec{},
 				Status: configv1.ClusterVersionStatus{
 					History: []configv1.UpdateHistory{},
+					Desired: configv1.Release{Version: tc.desiredVersionInCV},
 				},
 			}
 			if len(tc.currVersion) > 0 {
```
Comment (wking) [inline pkg/payload/precondition/clusterversion/upgradeable_test.go:142]: The line below this appends a `Completed` history entry.  I think we want to expand the test-case structure so we can declare whether each test-case is mid-update or not, so we can get both `Progressing=True` and `Progressing=False` cases to exercise the new check?

**Comment by petr-muller** [comment]
https://github.com/openshift/release/pull/57408 should fix the hypershift jobs

File: pkg/payload/precondition/clusterversion/upgradeable.go (hunk 2 above)
Comment (hongkailiu) [reply pkg/payload/precondition/clusterversion/upgradeable.go]: I thought we did not need to any change on the consumer side and `Progressing=True` in the conditions would just be taken care of by some existing code to make `Upgradable=False`.
The Reason check was for the non-blocking case for acceptedRisks.

Now, I recover the logic of generating `precondition.Error` with the latest patch.

File: pkg/payload/precondition/clusterversion/upgradeable_test.go (hunk 3 above)
Comment (hongkailiu) [reply pkg/payload/precondition/clusterversion/upgradeable_test.go:144]: The field `versionPartiallyUpgraded` in the case is to claim if the cluster is in the middle of a upgrade.
`Progressing=False` is also added in the new patch.

File: pkg/cvo/upgradeable.go [hunk 4]
```diff
@@ -263,6 +274,45 @@ func (check *clusterVersionOverridesUpgradeable) Check() *configv1.ClusterOperat
 	return cond
 }
 
+type upgradeInProgressUpgradeable struct {
```
Comment (hongkailiu) [inline pkg/cvo/upgradeable.go:277]: @wking 
I am thinking if it is a good choice to use un `Upgradeable` type here?
Non-nil return from the `Check()` function for all the existing `Upgradeable`s [leads to](https://github.com/openshift/cluster-version-operator/blob/e546515213c8681ca44c52f178401cd47ad07d11/pkg/cvo/upgradeable.go#L97) `Upgradeable=False`

But this new one does not. It depends on the 2nd upgrade request from the user and we only block it if it is minor level.
I feel pushing all the logic here to precondition is more reasonable.

File: pkg/cvo/upgradeable.go [hunk 5]
```diff
@@ -263,6 +274,45 @@ func (check *clusterVersionOverridesUpgradeable) Check() *configv1.ClusterOperat
...
+	name     string
+	cvLister configlistersv1.ClusterVersionLister
+}
+
+func (check *upgradeInProgressUpgradeable) Check() *configv1.ClusterOperatorStatusCondition {
+	cond := &configv1.ClusterOperatorStatusCondition{
+		Type:   configv1.OperatorProgressing,
```
Comment (hongkailiu) [inline pkg/cvo/upgradeable.go:284]: Does `Type=configv1.OperatorProgressing` confuse the `collectUpgradeableConditions` function?


https://github.com/openshift/cluster-version-operator/blob/3f49cddaf1d2dc58cb07e6a072909d9392b33554/pkg/cvo/upgradeable.go#L160

**Review by wking — APPROVED** [review]
/lgtm

**Review by wking — APPROVED** [review]
0366e77ea1 looks good to me too.

/lgtm

**Review by petr-muller — APPROVED** [review]

**Review by wking — APPROVED** [review]
/lgtm

**Review by petr-muller — APPROVED** [review]

**Comment by evakhoni** [comment]
pre-merge verified successfully in all four varients
/label qe-approved

**Comment by wking** [comment]
[`operators should not create watch channels very often`][1] is unrelated:

/override ci/prow/e2e-agnostic-ovn

[1]: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/1080/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-ovn/1861146112841224192

**Pull request merged**